src/higlass/widget.js

```diff
@@ -302,3 +324,76 @@ export default {
     };
   },
 };
+
+/**
+ * @template T
+ * @template U
+ * @typedef {{ data: T, resolve: (success: U) => void, reject: (err: unknown) => void }} WithResolvers
+ */
+
+/**
+ * Collects multiple calls within the same frame to processes them as a batch.
+ *
+ * The provided `processBatch` function receives an array of items, each with
+ * associated resolvers for fulfilling the original request.
+ *
+ * @template T
+ * @template U
+ * @param {(batch: Array<WithResolvers<T, U>>) => void} processBatch - A function to process each batch.
+ * @returns {(item: T) => Promise<U>} - A function to enqueue items.
+ */
+function consolidator(processBatch) {
+  /** @type {Array<WithResolvers<T, U>>} */
+  let pending = [];
+  /** @type {number} */
+  let id = 0;
+
+  function run() {
+    processBatch(pending);
+    pending = [];
+    id = 0;
+  }
+
+  return function enqueue(data) {
+    id = id || requestAnimationFrame(() => run());
+    let { promise, resolve, reject } = /** @type {typeof defer<U>} */ (defer)();
+    pending.push({ data, resolve, reject });
+    return promise;
+  };
+}
+
+/**
+ * @template T
+ * @returns {{ promise: Promise<T>, resolve: (success: T) => void, reject: (err: unknown) => void }}
+ */
+function defer() {
```

Comment:
I can't wrap my head around what this function is supposed to do.

Reply:
It's a polyfill fro something like https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers. 

Basically, it lets you create a "deferred" value. Normally, you'd resolve or reject a promise directly within the promise boy. But this popular technique lets you return the promise directly (for something else to "await") and then pass around the resolvers for that internally.

Reply:
Oh, actually `Promise.withResolvers` is pretty widely available now according to MDN. I think I'll just use it direcly.